internal/api/authorization/k8s.go

```diff
@@ -0,0 +1,639 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authorization
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"slices"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/perses/perses/internal/api/authorization/native"
+	"github.com/perses/perses/internal/api/utils"
+	"github.com/perses/perses/pkg/model/api/config"
+	v1 "github.com/perses/perses/pkg/model/api/v1"
+	v1Role "github.com/perses/perses/pkg/model/api/v1/role"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	authenticationclient "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	authorizationclient "k8s.io/client-go/kubernetes/typed/authorization/v1"
+
+	apiInterface "github.com/perses/perses/internal/api/interface"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/authenticatorfactory"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/authorization/authorizerfactory"
+	"k8s.io/apiserver/pkg/server/options"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func NewK8sAuthz(conf config.Config) (*k8sImpl, error) {
+	if !conf.Security.Authorization.Providers.Kubernetes.Enabled {
+		return nil, fmt.Errorf("kubernetes authorization is not enabled")
+	}
+	kubeconfig, err := InitKubeConfig(conf.Security.Authorization.Providers.Kubernetes.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Since the Authenticator will need to make a lot of requests to check all permissions, raise the
+	// default limits
+	kubeconfig.QPS = 500
+	kubeconfig.Burst = 1000
+
+	kubeClient, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return nil, nil
+	}
+
+	sarClient := kubeClient.AuthorizationV1()
+	authorizer, err := newAuthorizer(sarClient)
+	if err != nil {
+		return nil, nil
+	}
+
+	tokenClient := kubeClient.AuthenticationV1()
+	kubernetesAuthenticator, err := newAuthenticator(
+		tokenClient,
+	)
+	if err != nil {
+		return nil, nil
+	}
+
+	return &k8sImpl{
+		Authenticator:    kubernetesAuthenticator,
+		Authorizer:       authorizer,
+		Kubeconfig:       kubeconfig,
+		kubeClient:       kubeClient,
+		guestPermissions: conf.Security.Authorization.GuestPermissions,
+	}, nil
+}
+
+type k8sImpl struct {
+	Authenticator    authenticator.Request
+	Authorizer       authorizer.Authorizer
+	Kubeconfig       *rest.Config
+	kubeClient       *kubernetes.Clientset
+	guestPermissions []*v1Role.Permission
+}
+
+type K8sAction string
+
+const (
```

Comment:
for a better readability, I would put these constant, associated types and associated functions in another file.